dags/sustainment/jira_to_yaml.py

```diff
@@ -0,0 +1,290 @@
+                "notes": fields["description"]["content"][0]["content"][0]["text"]
+                if fields["description"]
+                else None,
+                "resources": {},
+            }
+            print(issue_metadata)
+            package_id = fields["summary"].lower().replace(" ", "-")
+
+            return {package_id: issue_metadata}
+        else:
+            logging.info("This is a request for updating exisitng dataset.")
+
+    def write_issue_content_to_yaml(issue_content, filename):
+        """Receives a json input and writes it to a YAML file"""
+
+        logging.info(f"Generating yaml file: {filename}")
+        try:
+            with open(YAML_DIR_PATH / filename, "w") as file:
+                yaml.dump(issue_content, file, sort_keys=False)
+        except PermissionError:
+            message = "Note: yaml file already exist, please double check!"
+            raise Exception(message)
+        return {"filename": filename}
+
+    def close_ticket(jira_issue_id):
+        url = (
+            "https://toronto.atlassian.net/rest/api/3/issue/"
+            + jira_issue_id
+            + "/transitions"
+        )
+        payload = json.dumps(
+            {"transition": {"id": jira_issue_transitions_mapping["In Progress"]}}
+        )
+
+        headers = {"Authorization": JIRA_API_KEY, "Content-Type": "application/json"}
+
+        response = requests.request("POST", url, headers=headers, data=payload)
+        logging.info(f"Status: {response.status_code}, Ticket has been closed.")
+
+    def get_all_jira_issues(**kwargs):
+        """Generate all yamls for "Publish Dataset" Jira issues"""
+
+        input_ticket_ids = kwargs.pop("ti").xcom_pull(task_ids="get_jira_ticket_id")[
+            "jira_ticket_id"
+        ]
+        logging.info(f"Jira Ticket Ids: {input_ticket_ids}")
+
+        # get all jira issues is and request type
+        issues = json.loads(requests.get(JIRA_URL, headers=headers).content)["issues"]
+        output_list = {}
+
+        for issue in issues:
+            if issue["key"] in input_ticket_ids:
+                request_type = issue["fields"]["customfield_10502"]["requestType"][
+                    "name"
+                ]
+                logging.info(
+                    f"Jira Ticket Id: {issue['key']} and Request Type: {request_type}"
+                )
+
+                issue_content = grab_issue_content(issue)
+                if issue_content:
+                    filename = "".join(list(issue_content.keys())) + ".yaml"
+                    # write data to yaml file
+                    write_issue_content_to_yaml(issue_content, filename)
+                    # close current ticket
+                    close_ticket(issue["key"])
+                    output_list[issue["key"]] = ":done_green:" + filename
+                else:
+                    output_list[
+                        issue["key"]
+                    ] = ("Request for updating existing dataset," +
```

Comment:
I thought that `metadata_generator.py`'s logic would live here.

I imagined when someone enters a Jira Ticket ID into `jira_to_yaml.py`, it creates a YAML for each ticket, regardless of whether it is a "Publish New Dataset" type ticket or a "Update Existing Dataset" type ticket.

This can be useful if someone wants to put a filestore dataset into the datastore.

Yanan let me know what you think of this, and if it's a lot of work to make this change. Let me know if you need to chat, or if something does not make sense :)

Reply:
I see your points mac.👍 This won't be a lot of work, just some adjustments of existing code. 

Some of considerations: 1) to what percentage an update existing request will need to create a yaml(upgrade to datastore), if many update request no need to upgrade to datastore, then we need to delete the generated yamls, but ofc delete is always easy than create.
2) for metadata_generator.py, we need to make sure the package name strictly consistent with the name of existing package on CKAN, so currently It's an input from Airflow UI and we assign it before running the dag. If we incorporate to jira_to_yaml, this could be an additional inputs parameters from UI together with Jira_ticket_id. We can not count on jira ticket "summary" field as package name(which we processed and then use as [yaml filename/package name] for Publish new dataset), since it's input from client, there is a great chance they won't use the exact package name.

What your thoughts?

Reply:
Good points :) Im glad you brought these up. 

My responses:

1. I would guess about half of the upgrade existing requests require new YAMLs. I think, if a YAML already exists, then this job *should not be able* update that YAML.

2. I think that we can use the _summary_ field reliably to get a package name from Jira, since clients tend to copy-paste the name from the browser. If, for some reason, they have misspelled it, we (Open Data) can correct the typo directly in Jira. I dont think we need to add another input parameter for this

Reply:
ok, I see. Thanks for the response. Current logic should raise an error if an yaml already exists. :)

Yeah, the **summary** field should be something like "Red Light Camera Annual Charges" can be simply processed to "red-light-camera-annual-charges" which matches the existing package on portal, and then use as package name for metadata_generator.py

I'll start incorporate the metadata_generator inside jira_to_yaml. I may create a "Update Existing Dataset"for testing. 👍

Reply:
Ohhh, right!! then this is easier. I can grab the package name from link! Thanks!